manipulation/src/export.cpp

```diff
@@ -0,0 +1,233 @@
+      at_type = at::kByte;
+      break;
+    case at::kShort:
+      onnx_type = onnx::TensorProto_DataType_INT16;
+      at_type = at::kShort;
+      break;
+    case at::kInt:
+      onnx_type = onnx::TensorProto_DataType_INT32;
+      at_type = at::kInt;
+      break;
+    case at::kLong:
+      onnx_type = onnx::TensorProto_DataType_INT64;
+      at_type = at::kLong;
+      break;
+    default:
+      torch::barf("unexpected tensor scalar type");
+      break;
+  }
+  p->set_data_type(onnx_type);
+  at::Tensor cont = tensor.toType(at::CPU(at_type)).contiguous();
+  p->set_raw_data(cont.data_ptr(), cont.type().elementSizeInBytes() * cont.numel());
+}
+
+void addAttribute(onnx::NodeProto * n_p, optimization::Node * n, optimization::Symbol name) {
+  auto attr = n_p->add_attribute();
+  attr->set_name(optimization::symbolToString(name));
+  switch(n->kindOf(name)) {
+    case AttributeKind::f:
+      attr->set_f(n->f(name));
+      attr->set_type(onnx::AttributeProto_AttributeType_FLOAT);
+      break;
+    case AttributeKind::fs:
+      attr->set_type(onnx::AttributeProto_AttributeType_FLOATS);
+      for(auto & v : n->fs(name))
+        attr->add_floats(v);
+      break;
+    case AttributeKind::i:
+      attr->set_type(onnx::AttributeProto_AttributeType_INT);
+      attr->set_i(n->i(name));
+      break;
+    case AttributeKind::is:
+      attr->set_type(onnx::AttributeProto_AttributeType_INTS);
+      for(auto & v : n->is(name))
+        attr->add_ints(v);
+      break;
+    case AttributeKind::s:
+      attr->set_type(onnx::AttributeProto_AttributeType_STRING);
+      attr->set_s(n->s(name));
+      break;
+    case AttributeKind::ss:
+      attr->set_type(onnx::AttributeProto_AttributeType_STRINGS);
+      for(auto & v : n->ss(name))
+        attr->add_strings(v);
+      break;
+    case AttributeKind::t: {
+      attr->set_type(onnx::AttributeProto_AttributeType_TENSOR);
+      auto t = attr->mutable_t();
+      encodeTensor(t, n->t(name));
+    } break;
+    case AttributeKind::ts:
+      attr->set_type(onnx::AttributeProto_AttributeType_TENSORS);
+      for(auto & v : n->ts(name)) {
+        auto t = attr->add_tensors();
+        encodeTensor(t, v);
+      }
+      break;
+    case AttributeKind::g: {
+      attr->set_type(onnx::AttributeProto_AttributeType_GRAPH);
+      auto g = attr->mutable_g();
+      encodeGraph(g, n->g(name));
+    } break;
+    case AttributeKind::gs:
+      attr->set_type(onnx::AttributeProto_AttributeType_GRAPHS);
+      for(auto & v : n->gs(name)) {
+        auto g = attr->add_graphs();
+        encodeGraph(g, v);
+      }
+      break;
+  }
+}
+
+const Dimension* dd = nullptr;
+
+void encodeTypeProtoTensorType(onnx::TypeProto_Tensor* tensor_type, Value* n) {
+  onnx::TensorShapeProto* shape = tensor_type->mutable_shape();
+  TensorType* node_type = n->type().get();
+  const std::vector<Dimension>& dims = node_type->sizes();
+  for (const Dimension& d : dims) {
+    dd = &d;
```

Comment:
not sure where it came from, but it's weird

Reply:
++